File: TourismAgency/AgencyNetworking/src/main/java/protobuffprotocol/ProtoUtils.java
package protobuffprotocol;

import domain.Reservation;
import domain.TravelAgent;
import domain.Trip;
import org.apache.logging.log4j.core.pattern.TextRenderer;

import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class ProtoUtils {
    public static AgencyProtobufs.AgencyRequest createLoginRequest(TravelAgent user){
        AgencyProtobufs.TravelAgent userDTO = AgencyProtobufs.TravelAgent.newBuilder().setUserId(user.getId()).setUsername(user.getUsername()).setPasswd(user.getPasswd()).build();
        AgencyProtobufs.AgencyRequest request = AgencyProtobufs.AgencyRequest.newBuilder()
                .setType(AgencyProtobufs.AgencyRequest.Type.LOGIN)
                .setUser(userDTO).build();
        return request;
    }

    public static AgencyProtobufs.AgencyRequest createLogoutRequest(TravelAgent user){
        AgencyProtobufs.TravelAgent userDTO = AgencyProtobufs.TravelAgent.newBuilder().setUserId(user.getId()).setUsername(user.getUsername()).setPasswd(user.getPasswd()).build();
        AgencyProtobufs.AgencyRequest request = AgencyProtobufs.AgencyRequest.newBuilder()
                .setType(AgencyProtobufs.AgencyRequest.Type.LOGOUT)
                .setUser(userDTO).build();
        return request;
    }

    public static AgencyProtobufs.AgencyRequest createGetAllTripsRequest(){
        AgencyProtobufs.AgencyRequest request = AgencyProtobufs.AgencyRequest.newBuilder()
                .setType(AgencyProtobufs.AgencyRequest.Type.GET_ALL_TRIPS).build();
        return request;
    }

    public static AgencyProtobufs.AgencyRequest createFilteredTripsRequest(String name, LocalTime minHour, LocalTime maxHour){
        AgencyProtobufs.Filter filterDTO = AgencyProtobufs.Filter.newBuilder().setPlace(name).setMaxHour(maxHour.getHour()).setMinHour(minHour.getHour()).build();

        AgencyProtobufs.AgencyRequest request = AgencyProtobufs.AgencyRequest.newBuilder()
                .setType(AgencyProtobufs.AgencyRequest.Type.TRIPS_BY_NAME_HOUR)
                .setFilterTrips(filterDTO).build();
        return request;
    }
    public static AgencyProtobufs.AgencyRequest createNewUserRequest(TravelAgent user){
        AgencyProtobufs.TravelAgent userDTO = AgencyProtobufs.TravelAgent.newBuilder().setUserId(user.getId()).setUsername(user.getUsername()).setPasswd(user.getPasswd()).build();
        AgencyProtobufs.AgencyRequest request = AgencyProtobufs.AgencyRequest.newBuilder()
                .setType(AgencyProtobufs.AgencyRequest.Type.ADD_USER)
                .setUser(userDTO).build();
        return request;
    }
    public static AgencyProtobufs.AgencyRequest createNewReservationRequest(Reservation reservation){
        AgencyProtobufs.Reservation resDTO = AgencyProtobufs.Reservation.newBuilder()
                .setClient(reservation.getClient())
                .setTelefon(reservation.getTelefon())
                .setTripID(reservation.getTripId().getId())
                .setPlace(reservation.getTripId().getPlace())
                .setAvTickets(reservation.getTripId().getFreeTickets())
                .setNrTick(reservation.getNrTick())
                .build();
        AgencyProtobufs.AgencyRequest request = AgencyProtobufs.AgencyRequest.newBuilder()
                .setType(AgencyProtobufs.AgencyRequest.Type.ADD_RESERVATION)
                .setReservation(resDTO).build();
        return request;
    }

    public static AgencyProtobufs.AgencyResponse createOkResponse(){
        AgencyProtobufs.AgencyResponse response=AgencyProtobufs.AgencyResponse.newBuilder()
                .setType(AgencyProtobufs.AgencyResponse.Type.OK).build();
        return response;
    }



    public static AgencyProtobufs.AgencyResponse createErrorResponse(String text){
        AgencyProtobufs.AgencyResponse response=AgencyProtobufs.AgencyResponse.newBuilder()
                .setType(AgencyProtobufs.AgencyResponse.Type.ERROR)
                .setError(text).build();
        return response;
    }

    public static AgencyProtobufs.AgencyResponse createGetAllTripsResponse(Iterable<Trip> trips){

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

        AgencyProtobufs.AgencyResponse.Builder response = AgencyProtobufs.AgencyResponse.newBuilder()
                .setType(AgencyProtobufs.AgencyResponse.Type.GET_ALL_TRIPS);
        for (Trip t : trips){
            AgencyProtobufs.Trip tripDTO = AgencyProtobufs.Trip.newBuilder()
                    .setTripId(t.getId())
                    .setPlace(t.getPlace())
                    .setTransport(t.getTransport())
                    .setDate(t.getDate().format(formatter))
                    .setPrice(t.getPrice())
                    .setNrTickets(t.getNrTickets())
                    .setFreeTickets(t.getFreeTickets())
                    .build();
            response.addTrips(tripDTO);
        }
        return response.build();

    }

    public static AgencyProtobufs.AgencyResponse createFilteredTripsResponse(Iterable<Trip> trips){

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

        AgencyProtobufs.AgencyResponse.Builder response = AgencyProtobufs.AgencyResponse.newBuilder()
                .setType(AgencyProtobufs.AgencyResponse.Type.TRIPS_FILTERED);
        for (Trip t : trips){
            AgencyProtobufs.Trip tripDTO = AgencyProtobufs.Trip.newBuilder()
                    .setTripId(t.getId())
                    .setPlace(t.getPlace())
                    .setTransport(t.getTransport())
                    .setDate(t.getDate().format(formatter))
                    .setPrice(t.getPrice())
                    .setNrTickets(t.getNrTickets())
                    .setFreeTickets(t.getFreeTickets())
                    .build();
            response.addTrips(tripDTO);
        }
        return response.build();

    }

    public static AgencyProtobufs.AgencyResponse createNewReservationResponse(Reservation reservation){
        AgencyProtobufs.Reservation resDTO = AgencyProtobufs.Reservation.newBuilder()
                .setClient(reservation.getClient())
                .setTelefon(reservation.getTelefon())
                .setTripID(reservation.getTripId().getId())
                .setPlace(reservation.getTripId().getPlace())
                .setAvTickets(reservation.getTripId().getFreeTickets())
                .setNrTick(reservation.getNrTick())
                .build();
        AgencyProtobufs.AgencyResponse response = AgencyProtobufs.AgencyResponse.newBuilder()
                .setType(AgencyProtobufs.AgencyResponse.Type.NEW_RESERVATION)
                .setReservation(resDTO).build();
        return response;
    }
   public static String getError(AgencyProtobufs.AgencyResponse response){
       String errorMessage=response.getError();
       return errorMessage;
   }
    public static TravelAgent getUser(AgencyProtobufs.AgencyResponse response){
        TravelAgent sender = new TravelAgent(response.getUser().getUsername(),response.getUser().getPasswd());
        sender.setId(response.getUser().getUserId());
        return sender;
    }

    public static Iterable<Trip> getTrips(AgencyProtobufs.AgencyResponse response){
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        List<Trip> trips = new ArrayList<>();
        for(int i=0;i<response.getTripsCount();i++){
            AgencyProtobufs.Trip tripDTO = response.getTrips(i);
            Trip trip = new Trip(tripDTO.getPlace(),tripDTO.getTransport(),LocalDateTime.parse(tripDTO.getDate(), formatter),tripDTO.getPrice(), tripDTO.getNrTickets(), tripDTO.getFreeTickets());
            trip.setId(tripDTO.getTripId());
            trips.add(trip);
        }
        return trips;
    }
    public static Reservation getReservation(AgencyProtobufs.AgencyResponse response){
        AgencyProtobufs.Reservation reservationDTO = response.getReservation();
        String client=reservationDTO.getClient();
        String telefon= reservationDTO.getTelefon();
        Trip trip=new Trip(reservationDTO.getTripID(), reservationDTO.getPlace(), reservationDTO.getAvTickets());
        Integer nrTick= reservationDTO.getNrTick();
        return new Reservation(client,telefon,trip,nrTick);
    }

    public static TravelAgent getUser(AgencyProtobufs.AgencyRequest request){
        TravelAgent sender = new TravelAgent(request.getUser().getUsername(),request.getUser().getPasswd());
        sender.setId(request.getUser().getUserId());
        return sender;
    }
    public static Reservation getReservation(AgencyProtobufs.AgencyRequest request){
        AgencyProtobufs.Reservation reservationDTO = request.getReservation();
        String client=reservationDTO.getClient();
        String telefon= reservationDTO.getTelefon();
        Trip trip=new Trip(reservationDTO.getTripID(), reservationDTO.getPlace(), reservationDTO.getAvTickets());
        Integer nrTick= reservationDTO.getNrTick();
        return new Reservation(client,telefon,trip,nrTick);
    }

}

File: AgentieTurism/protobuf_3/ProtoUtils.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using Agency.Protocol;
using model.domain;
using proto = Agency.Protocol;
using Trip = model.domain.Trip;

namespace protobuf_3
{
    static class ProtoUtils
    {
        public static AgencyRequest createLoginRequest(model.domain.TravelAgent user)
        {
            proto.TravelAgent userDTO = new proto.TravelAgent();
            userDTO.UserId = user.ID;
            userDTO.Username = user.Username;
            userDTO.Passwd = user.Passwd;

            AgencyRequest request = new AgencyRequest();
            request.Type = AgencyRequest.Types.Type.Login;
            request.User = userDTO;

            return request;
        }

        public static AgencyRequest createLogoutRequest(model.domain.TravelAgent user)
        {
            proto.TravelAgent userDTO = new proto.TravelAgent();
            userDTO.UserId = user.ID;
            userDTO.Username = user.Username;
            userDTO.Passwd = user.Passwd;

            AgencyRequest request = new AgencyRequest();
            request.Type = AgencyRequest.Types.Type.Logout;
            request.User = userDTO;

            return request;
        }

        public static AgencyRequest createNewUserRequest(model.domain.TravelAgent user)
        {
            proto.TravelAgent userDTO = new proto.TravelAgent();
            userDTO.UserId = user.ID;
            userDTO.Username = user.Username;
            userDTO.Passwd = user.Passwd;

            AgencyRequest request = new AgencyRequest();
            request.Type = AgencyRequest.Types.Type.AddUser;
            request.User = userDTO;

            return request;
        }

        public static AgencyRequest createNewReservation(model.domain.Reservation reservation)
        {
            proto.Reservation resDTO = new proto.Reservation();
            resDTO.Client = reservation.Client;
            resDTO.Telefon = reservation.Telefon;
            resDTO.Place = reservation.Trip.Place;
            resDTO.TripID = reservation.Trip.ID;
            resDTO.AvTickets = reservation.Trip.FreeTickets;
            resDTO.NrTick = reservation.NrTickets;

            AgencyRequest request = new AgencyRequest();
            request.Type = AgencyRequest.Types.Type.AddUser;
            request.Reservation = resDTO;
            return request;
        }

        public static AgencyRequest createGetAllTripsRequest()
        {
            AgencyRequest request = new AgencyRequest();
            request.Type = AgencyRequest.Types.Type.GetAllTrips;
            return request;
        }

        public static AgencyRequest createFilteredTripsRequest(String place, int minHour, int maxHour)
        {
            Filter filDTO = new Filter();
            filDTO.Place = place;
            filDTO.MinHour = minHour;
            filDTO.MaxHour = maxHour;

            AgencyRequest request = new AgencyRequest();
            request.Type = AgencyRequest.Types.Type.TripsByNameHour;
            request.FilterTrips = filDTO;
            return request;
        }

        public static AgencyResponse createOkResponse()
        {
            AgencyResponse response = new AgencyResponse {Type = AgencyResponse.Types.Type.Ok};
            return response;
        }


        public static AgencyResponse createErrorResponse(String text)
        {
            AgencyResponse response = new AgencyResponse
            {
                Type = AgencyResponse.Types.Type.Error, Error = text
            };
            return response;
        }

        public static AgencyResponse createGetAllTripsResponse(IEnumerable<model.domain.Trip> trips)
        {
            AgencyResponse response = new AgencyResponse();
            response.Type = AgencyResponse.Types.Type.GetAllTrips;
            foreach (var t in trips)
            {
                proto.Trip tripDTO = new proto.Trip();
                tripDTO.TripId = t.ID;
                tripDTO.Place = t.Place;
                tripDTO.Transport = t.Transport;
                tripDTO.Date = t.Date.ToString("yyyy-MM-dd HH:mm:ss");
                tripDTO.Price = t.Price;
                tripDTO.NrTickets = t.NrTickets;
                tripDTO.FreeTickets = t.FreeTickets;
                response.Trips.Add(tripDTO);
            }

            return response;
        }

        public static AgencyResponse createFilteredTripsResponse(IEnumerable<model.domain.Trip> trips)
        {
            AgencyResponse response = new AgencyResponse();
            response.Type = AgencyResponse.Types.Type.TripsFiltered;
            foreach (var t in trips)
            {
                proto.Trip tripDTO = new proto.Trip();
                tripDTO.TripId = t.ID;
                tripDTO.Place = t.Place;
                tripDTO.Transport = t.Transport;
                tripDTO.Date = t.Date.ToString("yyyy-MM-dd HH:mm:ss");
                tripDTO.Price = t.Price;
                tripDTO.NrTickets = t.NrTickets;
                tripDTO.FreeTickets = t.FreeTickets;
                response.Trips.Add(tripDTO);
            }

            return response;
        }

        public static AgencyResponse createNewReservationResponse(model.domain.Reservation reservation)
        {
            proto.Reservation resDTO = new proto.Reservation();
            resDTO.Client = reservation.Client;
            resDTO.Telefon = reservation.Telefon;
            resDTO.Place = reservation.Trip.Place;
            resDTO.TripID = reservation.Trip.ID;
            resDTO.AvTickets = reservation.Trip.FreeTickets;
            resDTO.NrTick = reservation.NrTickets;

            AgencyResponse response = new AgencyResponse();
            response.Type = AgencyResponse.Types.Type.NewReservation;
            response.Reservation = resDTO;
            return response;
        }
        public static String getError(AgencyResponse response)
        {
            String errorMessage = response.Error;
            return errorMessage;
        }
        public static model.domain.TravelAgent getUser(AgencyResponse response)
        {
            model.domain.TravelAgent user = new model.domain.TravelAgent(response.User.Username, response.User.Passwd);
            user.ID = response.User.UserId;
            return user;
        }

        public static model.domain.Reservation getReservation(AgencyResponse response)
        {
            proto.Reservation resDTO = response.Reservation;
            model.domain.Trip trip = new model.domain.Trip(resDTO.TripID, resDTO.Place, resDTO.AvTickets);
            model.domain.Reservation reservation =
                new model.domain.Reservation(resDTO.Client, resDTO.Telefon, trip, resDTO.NrTick);
            return reservation;
        }

        public static IEnumerable<model.domain.Trip> getTrips(AgencyResponse response)
        {
            List<model.domain.Trip> trips = new List<Trip>();
            for (int i = 0; i < response.Trips.Count; i++)
            {
                proto.Trip tripDTO = response.Trips[i];
                CultureInfo provider = CultureInfo.InvariantCulture;  
// It throws Argument null exception  
                DateTime dateTime = DateTime.ParseExact(tripDTO.Date, "yyyy-MM-dd HH:mm:ss", provider);
                model.domain.Trip trip = new model.domain.Trip(tripDTO.Place, tripDTO.Transport, dateTime,
                    tripDTO.Price, tripDTO.NrTickets, tripDTO.FreeTickets);
                trip.ID = tripDTO.TripId;
                
                trips.Add(trip);
            }

            return trips;
        }

        public static model.domain.TravelAgent getUser(AgencyRequest request)
        {
            model.domain.TravelAgent user = new model.domain.TravelAgent(request.User.Username, request.User.Passwd);
            user.ID = request.User.UserId;
            return user;
        }

        public static model.domain.Reservation getReservation(AgencyRequest request)
        {
            proto.Reservation resDTO = request.Reservation;
            model.domain.Trip trip = new model.domain.Trip(resDTO.TripID, resDTO.Place, resDTO.AvTickets);
            model.domain.Reservation reservation =
                new model.domain.Reservation(resDTO.Client, resDTO.Telefon, trip, resDTO.NrTick);
            return reservation;
        }
        
    }
}